A31-xds-routeaction-support.md

```diff
@@ -0,0 +1,226 @@
+A31: gRPC xDS RouteAction Support
+----
+* Author(s): Doug Fawley (dfawley)
+* Approver: markdroth, ejona86
+* Status: In Review
+* Implemented in:
+* Last updated: 2020-07-13
+* Discussion at: https://groups.google.com/g/grpc-io/c/iujkovEFSJU
+
+## Abstract
+
+gRPC provides per-method configuration for some features like timeouts and
+retries.  This configuration is part of the [Service
+Config](https://github.com/grpc/grpc/blob/master/doc/service_config.md),
+provided by the Name Resolver, and currently has limited matching semantics
+(per-service or per-method with exact name matching only).
+
+In xDS, RDS
+[provides](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/route/route_components.proto#route-routeaction)
+the ability to configure timeouts and several other features using its matchers,
+which support a more rich set of matching semantics than gRPC, including
+examining the outgoing HTTP headers and using regexes.
+
+Further complicating matters, gRPC needs to be able to determine timeout
+settings before invoking the Load Balancing Policy, which is the point at which
+the bulk of the current xDS support is integrated with gRPC.
+
+Extending gRPC to support these matchers directly is not a preferred solution to
+the problem, as it could impact performance for all gRPC users, increase our API
+surface, and grow our code base for features we otherwise do not intend to
+implement.
+
+## Proposal
+
+This proposal is broken up into two sections: the addition of new functionality
+to gRPC itself, and then a description of using that functionality to implement
+support for xDS RouteActions.
+
+### New Functionality in gRPC
+
+In order to allow the xDS matchers to configure gRPC behavior, but to avoid
+implementing support for them directly in gRPC, a new "per-RPC Config Selector"
+will be added for internal-only use.  This will be provided by the Name Resolver
+component when calling `Update` on the channel in a way that does not extend our
+publicly-visible API, similarly to how the xDS Client is provided to the Load
+Balancing Policy.
+
+If present, this Config Selector will be invoked once at the beginning of every
```

Comment:
We should probably mention what pieces of data are passed to the Config Selector. It is mentioned in the xDS portion below, but it seems appropriate to include here.